### PR TITLE
Added Davies-Bouldin Index and Dunn Index for KMEans Evaluation criteria

### PR DESCRIPTION
**Pull request by smarthi**

### CODE REVIEW

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  private static double calcDBIndex(Map<Integer, Tuple2<Double, Long>> clusterSumDistAndCounts,
+                                    List<ClusterInfo> clusters) {
+    double totalDBIndex = 0.0;
+    for (int i = 0; i < clusters.size(); i++) {
+      double maxDBIndex = 0;
+      for (int j = 0; j < clusters.size(); j++) {
+        if (i != j) {
+          ClusterInfo c1 = clusters.get(i);
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java]: This variable can be declared outside the inner loop, as well as scatter, and I think the `containsKey(i)` condition can be checked outside too? to avoid whole loops for `i` that won't do anything.

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      if (maxIntraClusterDistance < intraClusterDistance) {
+        maxIntraClusterDistance = intraClusterDistance;
+      }
+    }
+
+    double minInterClusterDistance = Double.POSITIVE_INFINITY;
+    for (int i = 0; i < clusters.size(); i++) {
+      // Distances are symmetric, hence d(i,j) == d(j,i)
+      for (ClusterInfo cluster : clusters) {
+        double distance = getDistance(clusters.get(i).getCenter(), cluster.getCenter());
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java]: Pretty minor here but getting the center could happen outside the inner loop. Is `i` needed in the outer loop or can it also be foreach?

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            double intraClusterDistanceSum = v1._1() + v2._1();
+            long clusteredPointsCount = v1._2() + v2._2();
+
+            return new Tuple2<>(intraClusterDistanceSum, clusteredPointsCount);
+          }
+        });
+  }
+
+  private static double getDistance(double[] v1, double[] v2) {
+    SquaredDistanceFn distanceFn = new SquaredDistanceFn();
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java]: I feel like this should be a member and this can be a non-static function rather than allocate an object every time.

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      final List<ClusterInfo> clusters, JavaRDD<Vector> testData) {
+
+    return testData.mapToPair(new PairFunction<Vector, Integer, Tuple2<Double, Long>>() {
+      @Override
+      public Tuple2<Integer, Tuple2<Double, Long>> call(Vector vector) {
+        double closestDist = Double.POSITIVE_INFINITY;
+        int minCluster = -1;
+        for (int i = 0; i < clusters.size(); i++) {
+          ClusterInfo cluster = clusters.get(i);
+          double distance = getDistance(cluster.getCenter(), vector.toArray());
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java]: `vector.toArray` can happen outside the loop. It's making a copy IIRC.

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java [hunk 5]
```diff
@@ -140,8 +142,24 @@ public double evaluate(JavaSparkContext sparkContext,
...
+
+    double cost;
+    switch (evaluationStrategy) {
+      case "DAVIES-BOULDIN" :
+        cost = Evaluation.daviesBouldinIndex(clusterInfoList, testingData);
+        break;
+      case "DUNN" :
+        cost = Evaluation.dunnIndex(clusterInfoList, testingData);
+        break;
+      default :
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java]: Let's name this case, like "SSE" or something, and let that be the default in the config file.

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java [hunk 6]
```diff
@@ -140,8 +142,24 @@ public double evaluate(JavaSparkContext sparkContext,
...
                          JavaRDD<String> testData) {
+    KMeansPMMLUtils.validatePMMLVsSchema(model, inputSchema);
     JavaRDD<Vector> testingData = parsedToVectorRDD(testData.map(MLFunctions.PARSE_FN));
-    double cost = pmmlToKMeansModel(model).computeCost(testingData.rdd());
+    List<ClusterInfo> clusterInfoList = KMeansPMMLUtils.read(model);
+
+    double cost;
+    switch (evaluationStrategy) {
+      case "DAVIES-BOULDIN" :
+        cost = Evaluation.daviesBouldinIndex(clusterInfoList, testingData);
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java]: This one's better when it's lower, no? we'd need something like 1 / index to make it larger when better. I am not sure what the range of this index is.

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+  static double daviesBouldinIndex(final List<ClusterInfo> clusters, JavaRDD<Vector> testData) {
+    Map<Integer, Tuple2<Double, Long>> clusterSumDistanceAndCounts =
+        fetchClusterSumDistanceAndCounts(clusters, testData).collectAsMap();
+
+    return calcDBIndex(clusterSumDistanceAndCounts, clusters);
+  }
+
+  /**
+   * Computes the Dunn Index of a given clustering, See http://en.wikipedia.org/wiki/Dunn_index
+   * @param clusters - List of {@link ClusterInfo}
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java]: Trivial - you can omit the " - " in javadoc as it won't quite print like this

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+   * @param testData - test data for evaluation
+   * @return Dunn Index - higher the better
+   */
+  static double dunnIndex(final List<ClusterInfo> clusters, JavaRDD<Vector> testData) {
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java]: Later we can add unit tests for these methods

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java [hunk 9]
```diff
@@ -59,10 +59,12 @@
   private final int numberOfRuns;
   private final List<HyperParamValues<?>> hyperParamValues;
   private final InputSchema inputSchema;
+  private final String evaluationStrategy;
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java]: Hm, i wonder if we should even use an enum for the evaluation strategy?

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java [hunk 10]
```diff
@@ -140,8 +142,24 @@ public double evaluate(JavaSparkContext sparkContext,
...
-    double cost = pmmlToKMeansModel(model).computeCost(testingData.rdd());
+    List<ClusterInfo> clusterInfoList = KMeansPMMLUtils.read(model);
+
+    double cost;
+    switch (evaluationStrategy) {
+      case "DAVIES-BOULDIN" :
+        cost = Evaluation.daviesBouldinIndex(clusterInfoList, testingData);
+        break;
+      case "DUNN" :
+        cost = Evaluation.dunnIndex(clusterInfoList, testingData);
```
Comment (srowen) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java]: Since this is higher when better, and it's inverted below, something has to be changed I think.

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java (hunk 9 above)
Comment (smarthi) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java]: Yes I was gonna make it an Enum.

File: oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java (hunk 6 above)
Comment (smarthi) [inline oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/KMeansUpdate.java]: doing that in the main Kmeansupdate.evaluate()

**Comment by srowen** [comment]
So there's a broader interesting question here. Since this is an unsupervised problem, there is no inherent need for test/train data. We could change the framework to also provide the training data set in the evaluate method, so that unsupervised methods can evaluate on test.union(train). I think it ultimately doesn't help to just evaluate on the test set, but, I admit I've never read up on this. Everything I recall evaluates on all of the data.

Another interesting issue here is that the clustering algorithm is not necessarily optimizing for the same thing being evaluated, but, I don't know if anything can be done about that now.

Yet another interesting issue is that this process is supposed to help choose k, but using SSE as an eval metric doesn't really help that. It will almost always tell you to choose a higher k. So yes we do need other eval metrics, but I haven't read through these to see if they have the same issue. Silhouette doesn't I think.
